src/Utilities/CssConcatenator/RuleBlock.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\Utilities\CssConcatenator;
+
+/**
+ * Implementation of media rules sub-property for \Pelago\Emogrifier\Utilities\CssConcatenator
+ *
+ * @internal
+ *
+ * @author Jake Hotson <jake.github@qzdesign.co.uk>
+ * @author SignpostMarv
+ */
+class RuleBlock
+{
+    /**
+     * Array whose keys are selectors for the rule block (values are of no
+     *     significance);
+     *
+     * @var mixed[]
+     *
+     * @psalm-var array<string, mixed>
+     */
+    public $selectorsAsKeys = [];
+
+    /**
+     * The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     *
+     * @var string
+     */
+    public $declarationsBlock;
```

Comment:
Please let's make the properties of the new classes `private` and use proper getters/setters so the properties are encapsulated.

Reply:
This might be the bias of having used psalm for so long, but why do the properties need to be encapsulated if static analysis ensures the properties are used appropriately?

Reply:
For several reasons (in addition to encapsulation of object properties having been a recognized best practice for a really long time):

- it allows us to change the internal representation of the data without having to change the API (and hence without changing the classes that use our class).
- It allows to have type safety during runtime (as arguments and return values can be typed, but fields cannot (yet)). Currently, any other class can write any data into the properties.
- It allows us to add logic to the getters or setters: additional validation, trimming, caching …

Reply:
> * it allows us to change the internal representation of the data without having to change the API

the classes are/should be flagged as internal

> * It allows to have type safety during runtime

in-repo code that would violate type safety during runtime would raise issues in psalm.

> * It allows us to add logic to the getters or setters: additional validation, trimming, caching …

defer getters/settters until additional logic is needed, as it is an internal class?

Reply:
> the classes are/should be flagged as internal

Even with internal classes, being able to update the internal representation withing having to change the API is a good thing.

@SignpostMarv  Another question: Do you have concerns with encapsulating the fields (i.e., you see potential pitfalls), or is it just not important to you? If it's the former, we'll need to address those concerns, and if it's the latter, then I'd prefer to spend my time implementing this than discussing something that actually is not a problem for anyone concerned.

Reply:
Then we follow different philosophies here: I think that unit tests, static analysis and manual code reviews complement each other, and that covering code with tests is paramount. (The optimal case is test-driven development.)

Reply:
> (And if I understand you correctly, you do not have any concerns, but just prefer another approach.)

yup- it's like with the tests that were identified as redundant- I'd be the type to remove those because static analysis says they're unneeded.

Reply:
> Then we follow different philosophies here: I think that unit tests, static analysis and manual code reviews complement each other, and that covering code with tests is paramount. (The optimal case is test-driven development.)

I'm with @oliverklee here.  Even though some tests and checks might overlap or even aim to do the same thing, they'll inevitably work in different ways, and have their own unique set of limitations.  If there's no overlap, things may fall through the gaps that you couldn't see :)

So I'd much rather have a huge overlap than the possibility of an tiny underlap which turns out to be gaping chasm.  It helps ensure software robustness.  A possible drawback from that approach is overloading Travis, but that's another issue.

BTW, it is good to discuss what approach/methodology is best, and question "are we doing things in the right way?".  Helps achieve perfection (which, philosophically, is never quite possible, but something we should always strive towards), and provides extracurricular things to think about.  I'm all for that :)

Reply:
"the best code is the code not written" :P